### ISSUE TITLE
Pickle example from Doc does not work

### ISSUE THREAD

**Issue by jczaja** [mpi4py/mpi4py#427]
Hi,

I'm fresh user of mpi4py , so perhaps I'm doing some obvious mistake.

I have a working multi-node MPI setup (two nodes) , where I can run MPI via C++ (locating MPI processes on different nodes)perfectly fine, but 
[basic pickle example](https://mpi4py.readthedocs.io/en/stable/tutorial.html#point-to-point-communication) (comm.recv and comm.send) does NOT work ONLY if MPI processes are located on different nodes (It works fine if MPI processes are located on the same node)

#### output:
```bash
File "mpi4py/MPI/Comm.pyx", line 1438, in mpi4py.MPI.Comm.recv
icas.sgi.com 1:   File "mpi4py/MPI/msgpickle.pxi", line 341, in mpi4py.MPI.PyMPI_recv
icas.sgi.com 1:   File "mpi4py/MPI/msgpickle.pxi", line 299, in mpi4py.MPI.PyMPI_recv_match
icas.sgi.com 1: mpi4py.MPI.Exception: Other MPI error
icas.sgi.com: rank 1 exited with code 1
```

Other examples from [tutorial](https://mpi4py.readthedocs.io/en/stable/tutorial.html#point-to-point-communication) (non-blocking communication and numpy arrays sending) work properly.

###### OS: SUSE Linux Enterprise Server 15 SP3

Please advise how to proceed,

**Comment by shssf** [comment]
As far as I see, It work if we use some useless temporal buffer `s_buf` like:
```
from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
s_buf = MPI.memory.allocate(42)

if rank == 0:
    send_data = [0,1,2,3]
    recv_data =comm.recv(s_buf, source=1)
    comm.send(send_data, dest=1)
elif rank == 1:
    send_data = [4,2,0,1]
    comm.send(send_data, dest=0)
    recv_data = comm.recv(s_buf, source=0)

```
But it stops working if we use 

`comm.recv(None, source=1)
`
If we have an issue at communication establishing stage - it should not work in both cases

**Comment by dalcinl** [comment]
@shssf I'm still waiting for a confirmation about whether you guys are using Intel MPI.

> As far as I see, It work if we use some useless temporal buffer `s_buf` like:

Yes, of course. If you provide a temporary buffer to receive the pickle stream, then mpi4py DOES NOT use `MPI_Mprobe`, and uses a regular `MPI_Recv` instead, as you can realize if you follow the logics in the [source code](https://github.com/mpi4py/mpi4py/blob/master/src/mpi4py/MPI/msgpickle.pxi#L394). The only caveat with this approach is that you have to know an upper-bound for the incoming message size corresponding to the pickled stream, and that is not always possible. But if you have such upper-bound, then providing this temporary buffer (which can be reused if you have to perform receives multiple times) not only fixes the issue with matched-probes, it is also faster, as internal memory allocations are no longer needed.

Setting `MPI4PY_RC_RECV_MPROBE=false` in the environment will also disable the use of `MPI_Mprobe`. However, this comes with a caveat: the implementation of `Comm.recv()` then has to resort to a regular `MPI_Probe` to query for the incoming message size and next issue the `MPI_Recv` call. This is obviously not thread-safe. And that's the reason the `MPI_Mprobe`-based implementation is the default one. If your application is not calling MPI point to point on the communicator in different threads, then it is totally OK to set `MPI4PY_RC_RECV_MPROBE=false`.

PS: I really wish Intel MPI folks could someday manage to fix this annoying issue. Intel MPI in the only MPI implementation I'm aware of where mpi4py doesn't "just work".

**Comment by shssf** [comment]
@dalcinl Hi, Perhaps I don't understand all details. As far as I understand, in the code line:
`recv_data =comm.recv(s_buf=None, source=1)`
Data should be received anyway but placed into `recv_data` variable.
I'm not sure if MPI_Mprobe used in this case.

Thank you for pointing a workaround. We will try this and report back about results.

**Comment by dalcinl** [comment]
@shssf Yes, if you use `recv(None, ...)`, then mpi4py will definitely use matched probes and receives. Things go like this: first, a call to `MPI_Mprove` blocks for an incoming message, next query the message size from `MPI_Status`, next internally allocate an auxiliary buffer, then receive the message with `MPI_Mrecv`, and finally unpickle the byte stream, such that the `recv_data` return value is (an unpickled copy of) the object that was sent. 
https://github.com/mpi4py/mpi4py/blob/master/src/mpi4py/MPI/msgpickle.pxi#L342-L363

The dispatch on the different ways to perform the `recv()` Python call are here:
https://github.com/mpi4py/mpi4py/blob/master/src/mpi4py/MPI/msgpickle.pxi#L394-L401
As you can see, if the `buf` argument is `None`, and the `options.recv_mprobe` flag is true (that's the default), then the `recv()` implementation based on `MPI_Mprobe` and `MPI_Mrecv` is the one used (`PyMPI_recv_match`).

PS: I'm still waiting for a confirmation about whether all this is happening with Intel MPI or a different MPI implementation.

**Comment by jczaja** [comment]
@dalcinl 

Thanks very much for writting detailed description of mechanics taking place in this process.

1. ```MPI4PY_RC_RECV_MPROBE=false```  works for us
2. As for MPI implementaion we are using  is based off of https://github.com/pmodels/mpich.
